### PR TITLE
preInvoke Middleware

### PR DESCRIPTION
**Pull request by AndrewRAM33**
preInvoke Middleware (#282)

`preInvoke` presents a confirmation or notification window that will be displayed before is executed `sendOperation` action.  `SendOperation` will be executed or not depend of type `preInvoke` action and user's actions.

Type of preInvoke action:
- `confirm` - window with text and yes / no buttons to confirm execution
- `info` - window with info message, `sendOperation` executed.
- `error` - window with error message, `sendOperation` not executed.

Features:
- Popup bottons custom text

Example preInvoke operation:
```
confirmOperation:
    message: 'Are you sure?'
    type: 'confirm'
```

### CODE REVIEW

File: src/components/widgets/TableWidget/TableWidget.tsx [hunk 1]
```diff
@@ -55,7 +55,7 @@ export interface TableWidgetProps extends TableWidgetOwnProps {
     filterGroups: FilterGroup[],
     onDrillDown: (widgetName: string, bcName: string, cursor: string, fieldKey: string) => void,
     onShowAll: (bcName: string, cursor: string, route: Route, widgetName: string) => void,
-    onOperationClick: (bcName: string, operationType: string, widgetName: string) => void,
+    onOperationClick: (bcName: string, operationType: string, widgetName: string, bcKey?: string, preInvoke?: OperationPreInvoke) => void,
```
Comment (Dergash) [inline src/components/widgets/TableWidget/TableWidget.tsx:58]: We should avoid complicating functions contract when possible.  
There are already a `widgetName` and `operationType` arguments, you can use them inside the middleware to find the widget, the operation and get `bcKey` and `preInvoke` from there

Comment (AndrewRAM33) [reply src/components/widgets/TableWidget/TableWidget.tsx]: fixed. Use from state.view.rowMeta

File: src/components/ui/Popup/Popup.tsx [hunk 2]
```diff
@@ -44,10 +46,10 @@ export const Popup: FunctionComponent<PopupProps> = (props) => {
                     }
                     <div className={styles.actions}>
                         <Button onClick={props.onOkHandler} className={styles.buttonYellow}>
-                            {t('Save')}
+                            {props.OkText ? props.OkText : t('Save')}
```
Comment (Dergash) [inline src/components/ui/Popup/Popup.tsx:49]:
```tsx
{props.OkText ?? t('Save')}
```

File: src/components/ui/Popup/Popup.tsx [hunk 3]
```diff
@@ -44,10 +46,10 @@ export const Popup: FunctionComponent<PopupProps> = (props) => {
...
                         </Button>
                         <Button onClick={props.onCancelHandler} className={styles.buttonCancel}>
-                            {t('Cancel')}
+                            {props.cancelText ? props.cancelText : t('Cancel')}
```
Comment (Dergash) [inline src/components/ui/Popup/Popup.tsx:52]:
```tsx
{props.cancelText ?? t('Cancel')}
```

File: src/middlewares/preInvokeMiddleware.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import {$do, types} from '../actions/actions'
+import {Store as CoreStore} from '../interfaces/store'
+
+const preInvokeAction = ({getState, dispatch}: MiddlewareAPI<Dispatch<AnyAction>, CoreStore>) => (next: Dispatch) =>
+    (action: AnyAction) => {
+        if (action.type === types.sendOperation) {
+            const bcName = action.payload.bcName
+            const operationType = action.payload.operationType
+            const widgetName = action.payload.widgetName
+            const preInvoke = action.payload?.confirmOperation
```
Comment (Dergash) [inline src/middlewares/preInvokeMiddleware.tsx:15]: `confirmOperation` field is deprecated:
https://github.com/tesler-platform/tesler-ui/blob/04f4874f83f6072ad938bdb749cb2d8cd5d03899/src/actions/actions.ts#L339-L342

IDE would show it if not for the missing typings; in some cases typescript can deduce the type based on the condition branch but not always: in this case it couldn't so you'll have to use manual case or write a typeguard to get correct typings for payload.

And with correct typings IDE will show that `confirmOperation` field is deprecated.
The reasoning is that when you need a pre-invoke you can calculate it based on `widgetName` and `operationType`: get the widget, read bcName, use bcName to access current row meta, use operationType to get the operation, read the `preInvoke` from there.  
The `confirmOperation` field is optional and intended for cases when you hardcode confirmation logic from front end.

File: src/middlewares/preInvokeMiddleware.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+            return preInvoke && !action.payload.confirm
```
Comment (Dergash) [inline src/middlewares/preInvokeMiddleware.tsx:17]: You probably do not need `action.payload.confirm` here

Comment (AndrewRAM33) [reply src/middlewares/preInvokeMiddleware.tsx:33]: It necessary that `preInvokeMiddleware`  not execute when confirm response is received

File: src/components/ui/Popup/Popup.tsx (hunk 2 above)
Comment (AndrewRAM33) [reply src/components/ui/Popup/Popup.tsx]: fixed

File: src/components/ui/Popup/Popup.tsx (hunk 3 above)
Comment (AndrewRAM33) [reply src/components/ui/Popup/Popup.tsx]: fixed

File: src/middlewares/preInvokeMiddleware.tsx (hunk 4 above)
Comment (AndrewRAM33) [reply src/middlewares/preInvokeMiddleware.tsx]: fixed. Get `preInvoke` Operation from state

**Pull request merged**